gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java

```diff
@@ -0,0 +1,191 @@
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageOptions;
+import com.google.cloud.hadoop.gcsio.StorageResourceId;
+import com.google.cloud.hadoop.gcsio.integration.GoogleCloudStorageTestHelper.TestBucketHelper;
+import com.google.cloud.hadoop.util.AsyncWriteChannelOptions;
+import com.google.cloud.storage.BlobWriteSessionConfigs;
+import com.google.common.collect.ImmutableSet;
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Iterator;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+
+/**
+ * Tests that are very specific to Java-storage client AND will nto make sense for Apiary client.
+ * Any generic test which is agnostic of client-type should reside in GoogleCloudStorageImplTest
+ */
+public class GoogleCloudStorageClientImplIntegrationTest {
+
+  private static final TestBucketHelper BUCKET_HELPER =
+      new TestBucketHelper("dataproc-gcs-client-impl");
+  private static final String TEST_BUCKET = BUCKET_HELPER.getUniqueBucketPrefix();
+
+  // Do cleanup the path after every test.
+  private static final String GCS_WRITE_TMP_DIR = "/tmp/gcs-write-dir";
+  private static final String GCS_WRITE_TMP_DIR_1 = "/tmp/gcs-write-dir-1";
+  /**
+   * Deafult value is picked from java-storage API.
+   *
+   * @see BlobWriteSessionConfigs#bufferToTempDirThenUpload()
+   */
+  private static final String DEFAULT_TEMP_SUB_DIR = "google-cloud-storage";
+
+  private static GoogleCloudStorage helperGcs;
+
+  @Rule
+  public TestName name =
+      new TestName() {
+        // With parametrization method name will get [index] appended in their name.
+        @Override
+        public String getMethodName() {
+          return super.getMethodName().replaceAll("[\\[,\\]]", "");
+        }
+      };
+
+  @BeforeClass
+  public static void before() throws IOException {
+
+    helperGcs = GoogleCloudStorageTestHelper.createGcsClientImpl();
+    helperGcs.createBucket(TEST_BUCKET);
+  }
+
+  @AfterClass
+  public static void after() throws IOException {
+    try {
+      BUCKET_HELPER.cleanup(helperGcs);
+    } finally {
+      helperGcs.close();
+    }
+  }
+
+  @Before
+  public void setUp() {
+    System.setProperty("java.io.tmpdir", GCS_WRITE_TMP_DIR);
+  }
+
+  @After
+  public void cleanUp() {
+    ImmutableSet<String> tempDirs = ImmutableSet.of(GCS_WRITE_TMP_DIR_1, GCS_WRITE_TMP_DIR);
+    Iterator<String> iterator = tempDirs.stream().iterator();
+    while (iterator.hasNext()) {
+      String filePath = iterator.next();
+      for (File file : new File(filePath).listFiles()) {
+        file.delete();
+      }
+    }
+  }
+
+  @Test
+  public void writeToDefaultPathThenUploadEnabled() throws IOException {
```

Comment:
should we test the disabled case? Also the multiple dirs case.

Reply:
For disabled we have to verify that it is not writing to the temp location, correct? i.e. existing test might be hitting the case, but it is not verifying.

Reply:
added